File: LungInsight-Frontend/src/api/predictions.ts
import { apiClient } from '@/lib/axios';
import { MOCK_MODE, mockDelay } from '@/lib/mock';
import { PLACEHOLDER_IMAGE } from '@/lib/placeholder';
import { generateMockPredictions, mockPredictionResult } from '@/mocks/data';
import type { Prediction, PredictionListResponse } from '@/types';

let mockStore: Prediction[] | null = null;
function getMockStore(): Prediction[] {
  if (!mockStore) mockStore = generateMockPredictions(24);
  return mockStore;
}

export async function createPrediction(file: File): Promise<Prediction> {
  if (MOCK_MODE) {
    await mockDelay(1200); // upload + inference both take real time; mimic that
    const result = mockPredictionResult();
    getMockStore().unshift(result);
    return result;
  }
  const formData = new FormData();
  formData.append('file', file);
  const { data } = await apiClient.post<Prediction>('/api/v1/prediction', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return data;
}

export async function getPrediction(id: string): Promise<Prediction> {
  if (MOCK_MODE) {
    await mockDelay();
    const found = getMockStore().find((p) => p.id === id);
    if (!found) throw new Error('Prediction not found.');
    return found;
  }
  const { data } = await apiClient.get<Prediction>(`/api/v1/prediction/${id}`);
  return data;
}

/** Fetches a viewable URL for the original uploaded X-ray. In Supabase
 * storage mode this is a time-limited signed URL (the bucket is private),
 * so don't cache it long-term -- refetch if it might have expired. */
export async function getPredictionImageUrl(id: string): Promise<string> {
  if (MOCK_MODE) {
    await mockDelay(150);
    // No real uploaded file in mock mode -- same placeholder chest X-ray
    // silhouette used everywhere else in mock data.
    return PLACEHOLDER_IMAGE;
  }
  const { data } = await apiClient.get<{ url: string }>(`/api/v1/prediction/${id}/image-url`);
  return data.url;
}

export async function listPredictions(page = 1, pageSize = 20): Promise<PredictionListResponse> {
  if (MOCK_MODE) {
    await mockDelay();
    const all = getMockStore();
    const start = (page - 1) * pageSize;
    return {
      items: all.slice(start, start + pageSize),
      total: all.length,
      page,
      page_size: pageSize,
    };
  }
  const { data } = await apiClient.get<PredictionListResponse>('/api/v1/predictions', {
    params: { page, page_size: pageSize },
  });
  return data;
}

File: LungInsight-Backend/app/api/routers/predictions.py
"""Prediction endpoints: create, get by id, list."""

from __future__ import annotations
import uuid

from fastapi import APIRouter, Depends, File, Query, UploadFile, status

from app.api.dependencies import get_prediction_service, CurrentUser
from app.schemas.prediction import PredictionResponse, PredictionListResponse
from app.services.prediction_service import PredictionService

router = APIRouter(prefix="/api/v1", tags=["Predictions"])


@router.post("/prediction", response_model=PredictionResponse, status_code=status.HTTP_201_CREATED)
async def create_prediction(
    current_user: CurrentUser,
    file: UploadFile = File(...),
    prediction_service: PredictionService = Depends(get_prediction_service),
) -> PredictionResponse:
    prediction = await prediction_service.create_prediction(current_user.id, file)
    return prediction


@router.get("/prediction/{prediction_id}", response_model=PredictionResponse)
def get_prediction(
    prediction_id: uuid.UUID,
    current_user: CurrentUser,
    prediction_service: PredictionService = Depends(get_prediction_service),
) -> PredictionResponse:
    return prediction_service.get_prediction(prediction_id, current_user.id)


@router.get("/prediction/{prediction_id}/image-url")
def get_prediction_image_url(
    prediction_id: uuid.UUID,
    current_user: CurrentUser,
    prediction_service: PredictionService = Depends(get_prediction_service),
) -> dict:
    """Returns a viewable URL for the original uploaded X-ray. In local
    storage mode this is a static file path served by this backend; in
    Supabase mode it's a time-limited signed URL (the bucket is private)."""
    url = prediction_service.get_image_url(prediction_id, current_user.id)
    return {"url": url}


@router.get("/predictions", response_model=PredictionListResponse)
def list_predictions(
    current_user: CurrentUser,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    prediction_service: PredictionService = Depends(get_prediction_service),
) -> PredictionListResponse:
    items, total = prediction_service.list_predictions(current_user.id, page, page_size)
    return PredictionListResponse(items=items, total=total, page=page, page_size=page_size)
